Add an option to hide title

kean opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

This PR adds an option to hide the title field.

<img width="320" alt="Screenshot 2025-02-13 at 11 18 42 AM" src="https://github.com/user-attachments/assets/df44127c-31f6-4417-871f-f8c48dbcbdb2" />

I also cleaned up `EditorViewController` a bit:

- Move parameters to a `EditorConfiguration` struct. It's easier to use, read, document, and evolve.
- Remove unused `EditorService`. There is still a it more code that could be removed.
- Remove unused `isEditorLoaded`.  

## Why?

For Comment Composer in jpios.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions

Pass the new option and verify that the title field is hidden:

```swift
        var configuration = EditorConfiguration()
        configuration.isTitleHidden = true

        let viewController = EditorViewController(configuration: configuration)
```

Verify that the editor still works when the settings are undefined (Android). 

### Accessibility Testing Instructions
<!-- How can you test the changes by using a keyboard/screen reader only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

kean committed:
Extract EditorConfiguration

kean committed:
Add isTitleHidden option

kean committed:
Remove isEditorLoaded

kean commented:
@dcalhoun, I'd appreciate a little bit of help with CSS. Without the title div, there is a jump when you focus:

https://github.com/user-attachments/assets/b3537ed3-df0f-4ca2-a1b4-235c8eb2489b

I also couldn't get the demo `dev-server` work in an iOS simulator. It used to work, but it's not loading anymore. Do you know anything about it?

dcalhoun commented:
> I'd appreciate a little bit of help with CSS. Without the title div, there is a jump when you focus:

I'll take a look at the jumping issue.  

> I also couldn't get the demo `dev-server` work in an iOS simulator. It used to work, but it's not loading anymore. Do you know anything about it?

It should work. Did you verify your Xcode [environment variable](https://github.com/wordpress-mobile/GutenbergKit/blob/4fc147c76ef790038859820cbf924ee3dbac9b03/README.md#ios) matches the IP address logged from the dev server? Are you able to load the dev server in your browser—both on your dev machine and in the simulator?

kean commented:
> id you verify your Xcode [environment variable](https://github.com/wordpress-mobile/GutenbergKit/blob/4fc147c76ef790038859820cbf924ee3dbac9b03/README.md#ios) matches the IP address logged from the dev server? 

Yes, did check that.

> Are you able to load the dev server in your browser—both on your dev machine and in the simulator?

It does load in my desktop browser and the browser in the simulator.

I was able to see the error after I removed `webView.alpha = 0`:

<img width="456" alt="Screenshot 2025-02-13 at 12 53 49 PM" src="https://github.com/user-attachments/assets/ad2a4b62-86d7-4a94-949d-7de9dbbdd7c9" />

```
@http://localhost:5173/components/visual-editor/use-editor-styles.js:31:26
__unstableMarkListeningStores@http://localhost:5173/@fs/Users/kean/Developer/GutenbergKit/node_modules/.vite/deps/chunk-M3BR7O3V.js:6483:27
updateValue@http://localhost:5173/@fs/Users/kean/Developer/GutenbergKit/node_modules/.vite/deps/chunk-M3BR7O3V.js:6916:63
@http://localhost:5173/@fs/Users/kean/Developer/GutenbergKit/node_modules/.vite/deps/chunk-M3BR7O3V.js:6948:16
_useMappingSelect@http://localhost:5173/@fs/Users/kean/Developer/GutenbergKit/node_modules/.vite/deps/chunk-M3BR7O3V.js:6967:12
useEditorStyles@http://localhost:5173/components/visual-editor/use-editor-styles.js:28:60
VisualEditor@http://localhost:5173/components/visual-editor/index.jsx:57:33
renderWithHooks@http://localhost:5173/@fs/Users/kean/Developer/GutenbergKit/node_modules/.vite/deps/chunk-674BCADQ.js:18474:35
mountIndeterminateComponent@http://localhost:5173/@fs/Users/kean/Developer/GutenbergKit/node_modules/.vite/deps/chunk-674BCADQ.js:21852:36
beginWork$1@http://localhost:5173/@fs/Users/kean/Developer/GutenbergKit/node_modules/.vite/deps/chunk-674BCADQ.js:26684:31
performUnitOfWork@http://localhost:5173/@fs/Users/kean/Developer/GutenbergKit/node_modules/.vite/deps/chunk-674BCADQ.js:26129:31
workLoopSync@http://localhost:5173/@fs/Users/kean/Developer/GutenbergKit/node_modules/.vite/deps/chunk-674BCADQ.js:26068:30
renderRootSync@http://localhost:5173/@fs/Users/kean/Developer/GutenbergKit/node_modules/.vite/deps/chunk-674BCADQ.js:26047:27
recoverFromConcurrentError@http://localhost:5173/@fs/Users/kean/Developer/GutenbergKit/node_modules/.vite/deps/chunk-674BCADQ.js:25664:42
performConcurrentWorkOnRoot@http://localhost:5173/@fs/Users/kean/Developer/GutenbergKit/node_modules/.vite/deps/chunk-674BCADQ.js:25612:56
workLoop@http://localhost:5173/@fs/Users/kean/Developer/GutenbergKit/node_modules/.vite/deps/chunk-674BCADQ.js:462:50
flushWork@http://localhost:5173/@fs/Users/kean/Developer/GutenbergKit/node_modules/.vite/deps/chunk-674BCADQ.js:441:30
performWorkUntilDeadline@http://localhost:5173/@fs/Users/kean/Developer/GutenbergKit/node_modules/.vite/deps/chunk-674BCADQ.js:649:50
```

It would be nice to be able to see it in the error and log it the terminal

dcalhoun commented:
That is odd that it loads in the browsers, but not the WebView. 

The error you shared seems to only be the stack trace. I recommend inspecting the WebView with Safari's developer tools to see if the actual error message is visible there.

kean commented:
The moment I inspected it using Safari, it started working, and is now works every time. It may be a coincidence. I'm not sure what was wrong, but its' probably something with my environment.

dcalhoun commented on ios/Sources/GutenbergKit/Sources/EditorConfiguration.swift:
```diff
@@ -0,0 +1,23 @@
+    /// The initial content to initialize the editor with.
+    public var content = ""
+
+    public var postID: Int?
+    public var postType: String?
+    public var themeStyles = false
+    public var plugins = false
+    public var isTitleHidden = false
+    public var siteURL = ""
+    public var siteApiRoot = ""
+    public var siteApiNamespace = ""
+    public var authHeader = ""
```
As we expand the use cases of GutenbergKit (GBK), making these optional seems sensible. 

In reality, though, the new `isTitleHidden`/`hideTitle` property should be unnecessary. The [post title](https://github.com/WordPress/gutenberg/blob/3df54ec6a39b342c20d8598524fd418155c55172/packages/editor/src/components/post-title/index.js#L199-L201) components includes support for disabling the title field based on the [post type's features](https://github.com/WordPress/gutenberg/blob/3df54ec6a39b342c20d8598524fd418155c55172/packages/editor/src/components/post-type-support-check/index.js). 

However, are these "comments" an actual post type/entity or merely a string of content passed to GBK? It begins presenting other questions that we should answer and/or address eventually: 

Are there presumptions in GBK that lead to errors when the optional configuration is absent? 

Should we export additional, simplified components for composing GBK in these types of contexts (comment editing)—e.g., separate `edit-post` and `editor` packages? The former might represent the current GBK that includes post type-related logic (post type fetching, site authentication, etc), the latter might represent a component used for simple block editing (comments).

kean replied:
> In reality, though, the new isTitleHidden/hideTitle property should be unnecessary. The [post title](https://github.com/WordPress/gutenberg/blob/3df54ec6a39b342c20d8598524fd418155c55172/packages/editor/src/components/post-title/index.js#L199-L201) components includes support for disabling the title field based on the [post type's features](https://github.com/WordPress/gutenberg/blob/3df54ec6a39b342c20d8598524fd418155c55172/packages/editor/src/components/post-type-support-check/index.js).

Excellent point. It's probably not "post type", as comments are separate entities and not CPTs, but it would make sense to resume the existing infrastructure like `PostTypeSupportCheck` if possible.

> Are there presumptions in GBK that lead to errors when the optional configuration is absent?

Right, it's not entirely clear what's required, and the fields that have to be provided together or not at all should be grouped in structs or enums.

> the latter might represent a component used for simple block editing (comments).

Composition generally works best. I'm afraid comments are more complicated than they initially seemed to me. To the very least, since they are connected to a site's media in P2, they need to know about your site. I'm sure there is more I'm unaware of: third-party blocks, for instance?

dcalhoun commented on ios/Sources/GutenbergKit/Sources/EditorViewController.swift:
```diff
@@ -6,58 +6,24 @@ import Combine
-    private var plugins: Bool
-    private var _initialRawContent: String
+
+    private let configuration: EditorConfiguration
```
Love the `EditorConfiguration` refactor. 👏🏻

dcalhoun reviewed (commented):
As you've noted, I opened #97 addressing some styling issues and improving the robustness of the option. 

It'd be ideal to mirror this feature on Android for parity, but I recognize it is not really necessary at the moment. I'm fine to postpone that for now. 

We need to implement this in `WordPress-iOS` before merging this work to avoid errors. Are you willing to do that please?

dcalhoun committed:
task: Expand hidden title robustness (#97)

* style: Format code

* feat: Apply opinionated default content styles

In the WordPress admin context, the block list receives "post content"
attributes that dictate the layout and alignment styles applied to the
blocks within the list. Currently, we cannot retrieve the post content
attributes via the REST API, as it does not include the current post
context.

Instead, we apply the "constrained" layout to improve the mobile editing
experience. However, this approach may ultimately introduce subtle
styling bugs--namely, classic vs block themes and editing posts/pages
that use non-traditional template layouts (e.g., full-width content).

* fix: Inject editor styles

Ensure various editor styles are injected. These styles are determined
by editor and theme settings. Unfortunately, we do not have all the
necessary context in the REST API to ensure all the styles are present.
Therefore, these changes also include some manual style additions that
are notably absent.

This manual approach is brittle. Ideally, the web editor relies upon the
same REST API endpoints as the mobile app rather than relying upon
server rendering context. This would push both platforms to ensure all
of the necessary settings are exposed.

* refactor: Pass renamed hideTitle setting as prop

Renamed for consistency with other settings. Passed as a prop to avoid
redundant bridge communication and unnecessary re-renders. Added option
to the text editor.

* docs: Document TextEditor

* refactor: Remove unused settings fetch

* docs: Document VisualEditor

* fix: Apply missing post title alignment styles

Apply similar constraints as found on the post content.

* fix: Address remote editor errors

Reinstate the layout component to simplify remote editor support.

* fix: Repair root level alignment conditional

The `themeStyles` setting is not utilized in the web editor and may be
unnecessary in this context.

* feat: Ensure reasonable defaults when theme styles are absent

This might be used when classic themes are active or in development
environments lack site/theme context.

* refactor: Hoist editor ready logic

Align with the web editor. Begin a path towards avoiding portions of the
editor UI appearing at different times.

kean committed:
Remove unused import

kean committed:
Build

kean commented:
I tested the changes from https://github.com/wordpress-mobile/GutenbergKit/pull/97 and merged it, rebased, and pushed a new build. Should be ready to merge.

dcalhoun commented:
> We need to implement this in WordPress-iOS before merging this work to avoid errors. Are you willing to do that please?

@kean following up on [my inquiry](https://github.com/wordpress-mobile/GutenbergKit/pull/94#pullrequestreview-2618230199) regarding WordPress-iOS. Do you intend to implement a matching WordPress-iOS PR?

kean commented:
The changes can be tested in https://github.com/wordpress-mobile/WordPress-iOS/pull/24089.

dcalhoun committed:
fix: Pad editor when theme styles are disabled

Avoid content flush against the device edge.

dcalhoun committed:
task: Capture build output

dcalhoun reviewed (approved):
Testing this worked well for me. I think this is ready to merge. 

As noted, in the future, we might consider [removing this hide title prop](https://github.com/wordpress-mobile/GutenbergKit/pull/94#discussion_r1956391974) or implementing [Android support](https://github.com/wordpress-mobile/GutenbergKit/pull/94#pullrequestreview-2618230199).
